File: c++-srcs/logic/gtest/LiteralTest.cc

/// @file LiteralTest.cc
/// @brief LiteralTest の実装ファイル
/// @author Yusuke Matsunaga (松永 裕介)
///
/// Copyright (C) 2017 Yusuke Matsunaga
/// All rights reserved.


#include "gtest/gtest.h"
#include "ym/Literal.h"


BEGIN_NAMESPACE_YM

TEST(LiteralTest, empty_constr)
{
  Literal lit;

  EXPECT_FALSE( lit.is_valid() );

  std::ostringstream oss;
  oss << lit;
  EXPECT_EQ( std::string("U<>"), oss.str() );
}

TEST(LiteralTest, simple_constr)
{
  SizeType var{3};
  Literal lit{var, false};

  EXPECT_TRUE( lit.is_valid() );

  auto var1 = lit.varid();
  EXPECT_EQ( var, var1 );

  EXPECT_TRUE( lit.is_positive() );
  EXPECT_FALSE( lit.is_negative() );

  auto lit_n = ~lit;
  EXPECT_NE( lit, lit_n );

  EXPECT_EQ( lit.varid(), lit_n.varid() );
  EXPECT_FALSE( lit_n.is_positive() );
  EXPECT_TRUE( lit_n.is_negative() );

  EXPECT_EQ( lit, lit.make_positive() );
  EXPECT_NE( lit, lit.make_negative() );

  EXPECT_NE( lit_n, lit_n.make_positive() );
  EXPECT_EQ( lit_n, lit_n.make_negative() );

  {
    std::ostringstream oss;
    oss << lit;
    EXPECT_EQ( std::string("P<3>"), oss.str() );
  }
  {
    std::ostringstream oss;
    oss << lit_n;
    EXPECT_EQ( std::string("N<3>"), oss.str() );
  }
}

TEST(LiteralTest, compare)
{
  SizeType var1{1};
  SizeType var2{2};
  Literal lit1p{var1, false};
  Literal lit1n{var1, true};
  Literal lit2p{var2, false};
  Literal lit2n{var2, true};

  // lit1p vs lit1p
  EXPECT_TRUE(  lit1p == lit1p );
  EXPECT_FALSE( lit1p != lit1p );
  EXPECT_FALSE( lit1p <  lit1p );
  EXPECT_FALSE( lit1p >  lit1p );
  EXPECT_TRUE(  lit1p <= lit1p );
  EXPECT_TRUE(  lit1p >= lit1p );

  // lit1p vs lit1n
  EXPECT_FALSE( lit1p == lit1n );
  EXPECT_TRUE(  lit1p != lit1n );
  EXPECT_TRUE( lit1p <  lit1n );
  EXPECT_FALSE(  lit1p >  lit1n );
  EXPECT_TRUE( lit1p <= lit1n );
  EXPECT_FALSE(  lit1p >= lit1n );

  // lit1n vs lit2p
  EXPECT_FALSE( lit1n == lit2p );
  EXPECT_TRUE(  lit1n != lit2p );
  EXPECT_TRUE(  lit1n <  lit2p );
  EXPECT_FALSE( lit1n >  lit2p );
  EXPECT_TRUE(  lit1n <= lit2p );
  EXPECT_FALSE( lit1n >= lit2p );

  // lit1p vs lit2n
  EXPECT_FALSE( lit1p == lit2n );
  EXPECT_TRUE(  lit1p != lit2n );
  EXPECT_TRUE(  lit1p <  lit2n );
  EXPECT_FALSE( lit1p >  lit2n );
  EXPECT_TRUE(  lit1p <= lit2n );
  EXPECT_FALSE( lit1p >= lit2n );

  // lit2p vs lit1n
  EXPECT_FALSE( lit2p == lit1n );
  EXPECT_TRUE(  lit2p != lit1n );
  EXPECT_FALSE( lit2p <  lit1n );
  EXPECT_TRUE(  lit2p >  lit1n );
  EXPECT_FALSE( lit2p <= lit1n );
  EXPECT_TRUE(  lit2p >= lit1n );
}

END_NAMESPACE_YM

File: py_capi/pytest/logic/literal_test.py
#! /usr/bin/env python3

"""Literal のテストプログラム

:file: literal_test.py
:author: Yusuke Matsunaga (松永 裕介)
:copyright: Copyright (C) 2023 Yusuke Matsunaga, All rights reserved.
"""

import pytest
from ymworks.logic import Literal


def test_constructor1():
    lit = Literal()

    assert not lit.is_valid()

    assert str(lit) == "U<>"

def test_constructor2():
    var0 = 0
    lit = Literal(var0)

    assert lit.is_valid()
    assert lit.is_positive()
    assert not lit.is_negative()
    assert lit.varid == var0

    assert str(lit) == "P<0>"

    lit_p = lit.make_positive()
    assert lit_p == lit

    lit_n = lit.make_negative()
    assert lit_n == ~lit


def test_constructor3():
    var0 = 1
    lit = Literal(var0, inv=False)

    assert lit.is_valid()
    assert lit.is_positive()
    assert not lit.is_negative()
    assert lit.varid == var0

    assert str(lit) == "P<1>"

    lit_p = lit.make_positive()
    assert lit_p == lit

    lit_n = lit.make_negative()
    assert lit_n == ~lit


def test_constructor4():
    var0 = 2
    lit = Literal(var0, inv=True)

    assert lit.is_valid()
    assert not lit.is_positive()
    assert lit.is_negative()
    assert lit.varid == var0

    assert str(lit) == "N<2>"

    lit_p = lit.make_positive()
    assert lit_p == ~lit

    lit_n = lit.make_negative()
    assert lit_n == lit


def test_bad_constructor1():
    with pytest.raises(Exception) as e:
        _ = Literal(0, True)
    assert e.type == TypeError
    assert str(e.value) == "function takes at most 1 positional argument (2 given)"


def test_compare1():
    lit1 = Literal(0, inv=False)
    lit2 = Literal(0, inv=True)

    assert not (lit1 == lit2)
    assert (lit1 != lit2)
    assert (lit1 < lit2)
    assert (lit1 <= lit2)
    assert not (lit1 > lit2)
    assert not (lit1 >= lit2)


def test_compare2():
    lit1 = Literal(0, inv=True)
    lit2 = Literal(1, inv=False)

    assert not (lit1 == lit2)
    assert (lit1 != lit2)
    assert (lit1 < lit2)
    assert (lit1 <= lit2)
    assert not (lit1 > lit2)
    assert not (lit1 >= lit2)


def test_compare3():
    lit1 = Literal(1, inv=True)
    lit2 = Literal(0, inv=False)

    assert not (lit1 == lit2)
    assert (lit1 != lit2)
    assert not (lit1 < lit2)
    assert not (lit1 <= lit2)
    assert (lit1 > lit2)
    assert (lit1 >= lit2)


def test_compare4():
    lit1 = Literal(1, inv=True)
    lit2 = Literal(1, inv=True)

    assert (lit1 == lit2)
    assert not (lit1 != lit2)
    assert not (lit1 < lit2)
    assert (lit1 <= lit2)
    assert not (lit1 > lit2)
    assert (lit1 >= lit2)
